driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -0,0 +1,437 @@
+        }
+
+        public void dispose() {
+            // nothing to do
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            clientNonce = randomStringGenerator.generate(RANDOM_LENGTH);
+
+            StringWriter writer = new StringWriter();
+            writer.append(GS2_HEADER);
+            writer.append("n=");
+            writer.append(getUserName());
+            writer.append(",r=");
+            writer.append(clientNonce);
+
+            String clientFirstMessage = writer.toString();
+            clientFirstMessageBare = clientFirstMessage.substring(3);
+            return decodeUTF8(clientFirstMessage);
+        }
+
+        private byte[] computeClientFinalMessage(final byte[] challenge) throws SaslException {
+            String serverFirstMessage = encodeUTF8(challenge);
+            HashMap<String, String> map = parseServerResponse(serverFirstMessage);
+            String serverNonce = map.get("r");
+            if (!serverNonce.startsWith(clientNonce)) {
+                throw new SaslException("Server sent an invalid nonce.");
+            }
+
+            StringWriter writer = new StringWriter();
+            writer.append("c=");
+            writer.append(encodeBase64(GS2_HEADER));
+            writer.append(",r=");
+            writer.append(serverNonce);
+
+            String clientFinalMessageWithoutProof = writer.toString();
+            String salt = map.get("s");
+            int iterationCount = Integer.parseInt(map.get("i"));
+            if (iterationCount < MINIMUM_ITERATION_COUNT) {
+                throw new SaslException("Invalid iteration count.");
+            }
+
+            String authMessage = clientFirstMessageBare + "," + serverFirstMessage + "," + clientFinalMessageWithoutProof;
+            writer.append(",p=");
+            writer.append(getClientProof(getAuthenicationHash(), salt, iterationCount, authMessage));
+            return decodeUTF8(writer.toString());
+        }
+
+        /**
+         * The client Proof:
+         * <p>
+         * AuthMessage     := client-first-message-bare + "," + server-first-message + "," + client-final-message-without-proof
+         * SaltedPassword  := Hi(Normalize(password), salt, i)
+         * ClientKey       := HMAC(SaltedPassword, "Client Key")
+         * ServerKey       := HMAC(SaltedPassword, "Server Key")
+         * StoredKey       := H(ClientKey)
+         * ClientSignature := HMAC(StoredKey, AuthMessage)
+         * ClientProof     := ClientKey XOR ClientSignature
+         * ServerSignature := HMAC(ServerKey, AuthMessage)
+         */
+        String getClientProof(final String password, final String salt, final int iterationCount, final String authMessage)
+                throws SaslException {
+            String hashedPassword = encodeUTF8(h(decodeUTF8(password + salt)));
+            String cacheKey = encodeBase64(h(decodeUTF8(hashedPassword + ":" + salt + ":" + iterationCount)));
```

Comment:
Why is hashedPassword digested twice?  And since it's digested at all, isn't there a possibility of collisions and therefore spurious failures?  I'm also uncomfortable with the use of ':' as a separator in the key, as that _could_ in weird circumstances result in a cache collision as well (if hashedPassword ends up containing colons).

How about instead use a class to represent the cache key?  Basically just a tuple of the password, salt, and count with equals/hashCode methods.

Reply:
Sounds good will add.